torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -481,24 +482,25 @@ def _root_pre_forward(
     )
     _clear_grads_if_needed(_all_handles(state))
 
+    # Prepares the forward inputs by moving them to ``compute_device``
+    # TODO: Do not use the side stream for tensor copies for now; investigate
+    # the perf with/without it.
+    args_tuple, kwargs_tuple = _to_kwargs(args, kwargs, state.compute_device.index, False)
+    args = args_tuple[0]
+    kwargs = kwargs_tuple[0]
 
-def _prepare_forward_inputs(
-    device: torch.device,
+    return args, kwargs
+
+
+def _cast_forward_inputs(
     input_dtype: Optional[torch.dtype],
     *args: Any,
     **kwargs: Any,
 ) -> Tuple[Any, Any]:
     """
-    Prepares the forward inputs by moving them to ``device`` and casting them
-    to ``input_dtype`` if it is not ``None``.
+    Prepares the forward inputs by casting them to ``input_dtype`` if it is not ``None``.
     """
-    # TODO: Do not use the side stream for tensor copies for now; investigate
-    # the perf with/without it.
-    # TODO: For mixed precision, move the inputs to the compute device and cast
-    # to reduced-precision in a single `to()` call.
-    args_tuple, kwargs_tuple = _to_kwargs(args, kwargs, device.index, False)
-    args = args_tuple[0]
-    kwargs = kwargs_tuple[0]
+    # TODO: For mixed precision, cast to reduced-precision in a single `to()` call.
```

Comment:
not specific to this PR, but we should definitely do this to avoid extra copying in mixed precision path.

Reply:
We should be careful about which stream we use. If we choose to create a new stream to do this, we should keep in mind the memory implications. Mixed precision casts are pretty cheap on GPU, so it may not be worth the implications on the caching allocator. I recommend doing some experiments if we want to make this change.